src/utils/sources-tree/addToTree.js

```diff
@@ -85,48 +87,52 @@ export function addToTree(
   url.path = decodeURIComponent(url.path);
 
   const parts = url.path.split("/").filter(p => p !== "");
-  const isDir = isDirectory(url);
   parts.unshift(url.group);
 
   let path = "";
+  let part = "";
   let subtree = tree;
+  let ancestor = tree;
 
   for (let i = 0; i < parts.length; i++) {
-    const part = parts[i];
+    part = parts[i];
     const isLastPart = i === parts.length - 1;
-
-    // Currently we assume that we are descending into a node with
-    // children. This will fail if a path has a directory named the
-    // same as another file, like `foo/bar.js/file.js`.
-    //
-    // TODO: Be smarter about this, which we'll probably do when we
-    // are smarter about folders and collapsing empty ones.
-
-    if (!nodeHasChildren(subtree)) {
-      return;
-    }
-
-    const children = subtree.contents;
-
-    let index = determineFileSortOrder(
-      children,
-      part,
-      isLastPart,
-      i === 0 ? debuggeeUrl : ""
-    );
-
-    const child = children.find(c => c.name === part);
-    if (child) {
-      // A node with the same name already exists, simply traverse
-      // into it.
-      subtree = child;
+    if (nodeHasChildren(subtree)) {
+      const child = subtree.contents.find(c => c.name === part);
+      if (child) {
+        // A node with the same name already exists, simply traverse
+        // into it. Keep track of the ancestor in case we need to make
+        // changes later on.
+        ancestor = subtree;
+        subtree = child;
+      } else {
+        // No node with this name exists, so insert a new one in the
+        // place that is alphabetically sorted.
+        const node = createNode(part, `${path}/${part}`, []);
+        let where = determineFileSortOrder(
+          subtree.contents,
+          part,
+          isLastPart,
+          i === 0 ? debuggeeUrl : ""
+        );
+
+        subtree.contents.splice(where, 0, node);
+        subtree = subtree.contents[where];
+      }
     } else {
-      // No node with this name exists, so insert a new one in the
-      // place that is alphabetically sorted.
-      const node = createNode(part, `${path}/${part}`, []);
-      const where = index === -1 ? children.length : index;
-      children.splice(where, 0, node);
-      subtree = children[where];
+      // we enter this else block if a file and a folder have the same
```

Comment:
at some point we might want to revisit this code, it feels like there is a simpler solution